crates/bitwarden-ipc/src/ipc_client.rs

```diff
@@ -42,170 +47,255 @@ pub struct IpcClientTypedSubscription<Crypto, Com, Ses, Payload>
+        };
+
+        #[cfg(not(target_arch = "wasm32"))]
+        tokio::spawn(future);
+
+        #[cfg(target_arch = "wasm32")]
+        wasm_bindgen_futures::spawn_local(future);
+    }
+
+    pub async fn is_running(self: &Arc<Self>) -> bool {
+        let incoming = self.incoming.read().await;
+        let cancellation_handle = self.cancellation_handle.read().await;
+        incoming.is_some() && cancellation_handle.is_some()
+    }
+
+    pub async fn stop(self: &Arc<Self>) {
+        let mut incoming = self.incoming.write().await;
+        let _ = incoming.take();
+
+        let mut cancellation_handle = self.cancellation_handle.write().await;
+        if let Some(cancellation_rx) = cancellation_handle.take() {
+            let _ = cancellation_rx.send(true);
+        }
+    }
+
     /// Send a message
-    pub async fn send(
-        self: &Arc<Self>,
-        message: OutgoingMessage,
-    ) -> Result<(), SendError<Crypto::SendError, Com::SendError>> {
-        self.crypto
+    pub async fn send(self: &Arc<Self>, message: OutgoingMessage) -> Result<(), Crypto::SendError> {
+        let result = self
+            .crypto
             .send(&self.communication, &self.sessions, message)
-            .await
+            .await;
+
+        if result.is_err() {
+            log::error!("Error sending message: {:?}", result);
+            self.stop().await;
+        }
+
+        result
     }
 
     /// Create a subscription to receive messages, optionally filtered by topic.
     /// Setting the topic to `None` will receive all messages.
     pub async fn subscribe(
         self: &Arc<Self>,
         topic: Option<String>,
-    ) -> IpcClientSubscription<Crypto, Com, Ses> {
-        IpcClientSubscription {
-            receiver: self.communication.subscribe().await,
+    ) -> Result<IpcClientSubscription<Crypto, Com, Ses>, SubscribeError> {
+        Ok(IpcClientSubscription {
+            receiver: self
+                .incoming
+                .read()
+                .await
+                .as_ref()
+                .ok_or(SubscribeError::NotStarted)?
+                .resubscribe(),
             client: self.clone(),
             topic,
-        }
+        })
     }
 
     /// Create a subscription to receive messages that can be deserialized into the provided payload
     /// type.
     pub async fn subscribe_typed<Payload>(
         self: &Arc<Self>,
-    ) -> IpcClientTypedSubscription<Crypto, Com, Ses, Payload>
+    ) -> Result<IpcClientTypedSubscription<Crypto, Com, Ses, Payload>, SubscribeError>
     where
         Payload: TryFrom<Vec<u8>> + PayloadTypeName,
     {
-        IpcClientTypedSubscription {
-            receiver: self.communication.subscribe().await,
+        Ok(IpcClientTypedSubscription {
+            receiver: self
+                .incoming
+                .read()
+                .await
+                .as_ref()
+                .ok_or(SubscribeError::NotStarted)?
+                .resubscribe(),
```

Comment:
Non-blocking thought: Some of the code for `IpcClientTypedSubscription` is fairly similar to the code in `IpcClientSubscription`, I wonder if we could define it like `IpcClientTypedSubscription<T>(IpcClientSubscription, PhantomData<T>)` or something like that to avoid repetition in their subscribe and receive functions.